File: examples/basic_agent.rs
use liteagent::llm::openai::OpenAiModel;
use liteagent::tool::FunctionTool;
use liteagent::Agent;
use serde_json::json;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    dotenvy::dotenv().ok();
    let api_key = std::env::var("OPENAI_API_KEY").expect("set OPENAI_API_KEY");
    let model = OpenAiModel::new("gpt-5-nano", api_key);

    let weather_tool = FunctionTool::new(
        "get_weather",
        "Get the current weather for a given city",
        json!({
            "type": "object",
            "properties": { "city": { "type": "string", "description": "City name" } },
            "required": ["city"]
        }),
        |args| {
            Box::pin(async move {
                let city = args["city"].as_str().unwrap_or("unknown").to_string();
                // Pretend this hits a real weather API.
                Ok(format!("It's 22C and sunny in {city}."))
            })
        },
    );

    let mut agent = Agent::builder(model)
        .instructions("You are a helpful weather assistant. Use tools when needed.")
        .tool(weather_tool)
        .build();

    let answer = agent.run("explain mixture of expert architecture?").await?;
    println!("{answer}");

    // Follow-up in the same in-memory conversation.
    let answer2 = agent.run("And how about Paris?").await?;
    println!("{answer2}");

    Ok(())
}

File: python-wrapper/examples/basic_agent.py
import asyncio
import os
from ferragent import Agent, Tool

def weather(args: dict) -> str:
    return f"It's 22C and sunny in {args.get('city', 'unknown')}."

async def main() -> None:
    tool = Tool("get_weather", "Get weather for a city", {
        "type": "object", "properties": {"city": {"type": "string"}}, "required": ["city"]
    }, weather)
    agent = Agent.openai("gpt-5-nano", os.environ["OPENAI_API_KEY"],
                         instructions="Use tools when needed.", tools=[tool])
    print(await agent.run("What's the weather in Bengaluru?"))
    print(await agent.run("And Paris?"))

asyncio.run(main())
